internal/cmd/commands/ferry/addtoken.go

```diff
@@ -0,0 +1,187 @@
+	})
+
+	f.StringVar(&base.StringVar{
+		Name:   "token",
+		Target: &c.FlagToken,
+		Usage:  `A URL pointing to a file on disk (file://) from which a token will be read or an env var (env://) from which the token will be read. Overrides the "token-name" parameter.`,
+	})
+
+	f.BoolVar(&base.BoolVar{
+		Name:   "output-curl-string",
+		Target: &c.FlagOutputCurlString,
+		Usage:  "Instead of executing the request, print an equivalent cURL command string and exit.",
+	})
+
+	f.UintVar(&base.UintVar{
+		Name:    "ferry-port",
+		Target:  &c.FlagFerryDaemonPort,
+		Default: 9300,
+		EnvVar:  base.EnvFerryDaemonPort,
+		Usage:   "The port on which the ferry daemon is listening.",
+	})
+
+	return set
+}
+
+func (c *AddTokenCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictNothing
+}
+
+func (c *AddTokenCommand) AutocompleteFlags() complete.Flags {
+	return c.Flags().Completions()
+}
+
+func (c *AddTokenCommand) Run(args []string) int {
+	ctx := c.Context
+	f := c.Flags()
+	if err := f.Parse(args); err != nil {
+		c.PrintCliError(err)
+		return base.CommandUserError
+	}
+	client, err := c.Client()
+	if err != nil {
+		c.PrintCliError(err)
+		return base.CommandCliError
+	}
+
+	resp, apiErr, err := c.Add(ctx, c.UI, client)
+	if err != nil {
+		c.PrintCliError(err)
+		return base.CommandCliError
+	}
+	if apiErr != nil {
+		c.PrintApiError(apiErr, "Error from daemon when adding a token")
+		return base.CommandApiError
+	}
+	switch base.Format(c.UI) {
+	case "json":
+		if ok := c.PrintJsonItem(resp); !ok {
+			return base.CommandCliError
+		}
+	case "table":
+		c.UI.Output("The ferry add-token operation completed successfully.")
+	}
+	return base.CommandSuccess
+}
+
+// userTokenToAdd is the request body to this handler.
+type UpsertTokenRequest struct {
+	// BoundaryAddr is a required field for all requests
+	BoundaryUrl string `json:"boundary_url,omitempty"`
+	// The raw auth token for this user.
+	Token string `json:"token,omitempty"`
+}
+
+// Add builds the UpsertTokenRequest using the client's address and token.
+// It then sends the request to the ferry daemon.
+// The passed in cli.Ui is used to print out any errors when looking up the
+// auth token from the keyring. This allows background operations calling this
+// method to pass in a silent UI to suppress any output.
+func (c *AddTokenCommand) Add(ctx context.Context, ui cli.Ui, apiClient *api.Client) (*api.Response, *api.Error, error) {
+	pa := UpsertTokenRequest{
+		BoundaryUrl: apiClient.Addr(),
+	}
+	token := apiClient.Token()
+	if token == "" {
+		return nil, nil, errors.New("The client auth token is empty.")
+	}
+	if parts := strings.SplitN(token, "_", 4); len(parts) != 3 {
```

Comment:
Why 4? I think this should be 3? The format we expect is `at_1234567890_awduawodnawodinaw21en1o23h1823h1n` or something like that, right? So we want to split this into at most 3 substrings.

Reply:
Ah, so you want to use `SplitN` where N=number of expected sections+1 because `Split` is potentially more expensive? I get it but I think it obscures the intent. How about just using `Split`?